File: heartbeat/src/heartbeat/signals/firestore_tokens.py
"""Phase F — Read encrypted OAuth tokens from Firestore.

Replaces the Phase E pattern of reading a static google-token.json.
Tokens are written by the Tauri app (AES-256-GCM encrypted) to
``engagements/{eid}/google_tokens/google`` and read + decrypted here.

If the access token is expired, this module refreshes it via the
Google token endpoint and writes the updated encrypted payload back
to Firestore (optimistic-concurrency: re-read doc after refresh,
skip writeback if another writer updated it in between).

Falls back to the legacy file-based path if Firestore token is
unavailable (backwards-compatible with Phase E deployments).
"""

from __future__ import annotations

import base64
import json
import logging
import os
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from heartbeat.signals.base import CollectorError
from heartbeat.signals.google_auth import (
    GoogleAuthFailure,
    GoogleSource,
    load_google_credentials,
)

if TYPE_CHECKING:
    from pathlib import Path

    from google.oauth2.credentials import Credentials

logger = logging.getLogger("heartbeat.signals.firestore_tokens")

# Cached Firestore client, initialized lazily via _get_db().
# typed as Any because google-cloud-firestore is untyped.
_FS_CLIENT: Any | None = None


def _get_db() -> Any:
    """Return a Firestore client, initializing the Firebase app if needed.

    Uses the same SA path as the outputs layer (FIREBASE_SA_KEY_PATH env var)
    but with a dedicated named app to avoid collisions.
    """
    global _FS_CLIENT  # noqa: PLW0603
    if _FS_CLIENT is not None:
        return _FS_CLIENT

    import firebase_admin
    from firebase_admin import credentials
    from firebase_admin import firestore as fs

    sa_path = os.environ.get("FIREBASE_SA_KEY_PATH", "/etc/ikrs-heartbeat/firebase-sa.json")
    app_name = "heartbeat-token-sync"
    try:
        app = firebase_admin.get_app(app_name)  # type: ignore[no-untyped-call]
    except ValueError:
        cred = credentials.Certificate(sa_path)  # type: ignore[no-untyped-call]
        app = firebase_admin.initialize_app(cred, name=app_name)  # type: ignore[no-untyped-call]

    _FS_CLIENT = fs.client(app=app)
    return _FS_CLIENT


@dataclass(frozen=True)
class TokenPayload:
    """Mirrors the Tauri TokenPayload struct."""

    access_token: str
    refresh_token: str
    expires_at: int
    client_id: str
    client_secret: str


def _get_encryption_key() -> tuple[bytes, int] | None:
    """Read the base64-encoded AES key + version from env.

    Returns (key_bytes, version) or None if not configured.
    """
    key_b64 = os.environ.get("TOKEN_ENCRYPTION_KEY", "")
    if not key_b64:
        return None
    try:
        key_bytes = base64.b64decode(key_b64)
    except Exception:
        logger.warning("TOKEN_ENCRYPTION_KEY is not valid base64")
        return None
    if len(key_bytes) != 32:
        logger.warning(
            "TOKEN_ENCRYPTION_KEY decoded to %d bytes, expected 32", len(key_bytes)
        )
        return None
    version = int(os.environ.get("TOKEN_ENCRYPTION_KEY_VERSION", "1"))
    return key_bytes, version


def _get_prev_encryption_key() -> tuple[bytes, int] | None:
    """Read the previous key for rotation fallback."""
    key_b64 = os.environ.get("TOKEN_ENCRYPTION_KEY_PREV", "")
    if not key_b64:
        return None
    try:
        key_bytes = base64.b64decode(key_b64)
    except Exception:
        return None
    if len(key_bytes) != 32:
        return None
    version = int(os.environ.get("TOKEN_ENCRYPTION_KEY_PREV_VERSION", "0"))
    return key_bytes, version


def _decrypt(ciphertext_b64: str, iv_b64: str, key_bytes: bytes) -> str:
    """Decrypt AES-256-GCM ciphertext. Returns plaintext string.

    The ciphertext_b64 contains ciphertext || 16-byte GCM auth tag
    (concatenated), matching WebCrypto's default output.
    """
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    ct = base64.b64decode(ciphertext_b64)
    iv = base64.b64decode(iv_b64)
    aesgcm = AESGCM(key_bytes)
    plaintext = aesgcm.decrypt(iv, ct, None)
    return plaintext.decode("utf-8")


def _encrypt(plaintext: str, key_bytes: bytes) -> tuple[str, str]:
    """Encrypt with AES-256-GCM. Returns (ciphertext_b64, iv_b64).

    ciphertext_b64 contains ciphertext || 16-byte GCM auth tag.
    """
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    iv = os.urandom(12)
    aesgcm = AESGCM(key_bytes)
    ct = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
    return base64.b64encode(ct).decode("ascii"), base64.b64encode(iv).decode("ascii")


def _read_firestore_token(
    engagement_id: str,
    source: GoogleSource,
) -> tuple[TokenPayload, dict[str, object]] | None:
    """Read and decrypt token from Firestore.

    Returns (payload, raw_doc_dict) or None if unavailable.
    Raises GoogleAuthFailure on decrypt/key errors.
    """
    key_info = _get_encryption_key()
    if key_info is None:
        return None  # Key not configured — fall back to legacy

    key_bytes, key_version = key_info

    db = _get_db()
    ref = db.document(f"engagements/{engagement_id}/google_tokens/google")
    snap = ref.get()

    if not snap.exists:
        return None  # No Firestore token — fall back to legacy

    doc = snap.to_dict() or {}
    ct_b64 = doc.get("ciphertext", "")
    iv_b64 = doc.get("iv", "")
    doc_key_version = doc.get("keyVersion", 1)

    if not ct_b64 or not iv_b64:
        logger.warning("Firestore token doc exists but is empty for engagement %s", engagement_id)
        return None

    # Try decryption with the matching key version
    decrypt_key: bytes | None = None
    if doc_key_version == key_version:
        decrypt_key = key_bytes
    else:
        prev = _get_prev_encryption_key()
        if prev is not None and doc_key_version == prev[1]:
            decrypt_key = prev[0]

    if decrypt_key is None:
        raise GoogleAuthFailure(
            CollectorError(
                source=source,
                error_code="key_version_unknown",
                message=(
                    f"Firestore token has keyVersion={doc_key_version} but "
                    f"local key is version {key_version}. Update the "
                    "encryption key on this machine."
                ),
            )
        )

    try:
        plaintext = _decrypt(ct_b64, iv_b64, decrypt_key)
    except Exception as exc:
        raise GoogleAuthFailure(
            CollectorError(
                source=source,
                error_code="token_decrypt_failed",
                message=f"Failed to decrypt Firestore token: {exc}",
            )
        ) from exc

    try:
        data = json.loads(plaintext)
    except json.JSONDecodeError as exc:
        raise GoogleAuthFailure(
            CollectorError(
                source=source,
                error_code="token_decrypt_failed",
                message=f"Decrypted token is not valid JSON: {exc}",
            )
        ) from exc

    payload = TokenPayload(
        access_token=data.get("access_token", ""),
        refresh_token=data.get("refresh_token", ""),
        expires_at=int(data.get("expires_at", 0)),
        client_id=data.get("client_id", ""),
        client_secret=data.get("client_secret", ""),
    )

    return payload, doc


def _is_expired(payload: TokenPayload) -> bool:
    """Check if access token is expired (with 5-min buffer)."""
    return payload.expires_at <= int(time.time()) + 300


def _refresh_and_writeback(
    payload: TokenPayload,
    engagement_id: str,
    original_doc: dict[str, object],
) -> TokenPayload:
    """Refresh the access token and write back to Firestore.

    Implements optimistic concurrency: re-reads the doc after refresh,
    skips writeback if updatedAt changed (another writer got there first).
    """
    import requests

    resp = requests.post(  # type: ignore[attr-defined,no-untyped-call]
        "https://oauth2.googleapis.com/token",
        data={
            "client_id": payload.client_id,
            "client_secret": payload.client_secret,
            "refresh_token": payload.refresh_token,
            "grant_type": "refresh_token",
        },
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()

    new_access_token = data["access_token"]
    new_expires_in = data.get("expires_in", 3600)
    # Google may rotate refresh_token — use new one if provided
    new_refresh_token = data.get("refresh_token", "") or payload.refresh_token

    refreshed = TokenPayload(
        access_token=new_access_token,
        refresh_token=new_refresh_token,
        expires_at=int(time.time()) + new_expires_in,
        client_id=payload.client_id,
        client_secret=payload.client_secret,
    )

    # Write back to Firestore with optimistic concurrency
    key_info = _get_encryption_key()
    if key_info is not None:
        key_bytes, key_version = key_info
        plaintext = json.dumps({
            "access_token": refreshed.access_token,
            "refresh_token": refreshed.refresh_token,
            "expires_at": refreshed.expires_at,
            "client_id": refreshed.client_id,
            "client_secret": refreshed.client_secret,
        })
        ct_b64, iv_b64 = _encrypt(plaintext, key_bytes)

        from google.cloud.firestore import SERVER_TIMESTAMP

        db = _get_db()
        ref = db.document(f"engagements/{engagement_id}/google_tokens/google")

        # Optimistic concurrency: re-read, skip if doc changed
        current = ref.get()
        if current.exists:
            current_updated = (current.to_dict() or {}).get("updatedAt")
            original_updated = original_doc.get("updatedAt")
            if current_updated != original_updated:
                logger.info(
                    "Firestore token updated by another writer; skipping writeback"
                )
                return refreshed

        ref.set({
            "ciphertext": ct_b64,
            "iv": iv_b64,
            "keyVersion": key_version,
            "updatedAt": SERVER_TIMESTAMP,
            "writtenBy": "heartbeat",
        })
        logger.info("Wrote refreshed token back to Firestore for engagement %s", engagement_id)

    return refreshed


def _payload_to_credentials(payload: TokenPayload) -> Credentials:
    """Convert a TokenPayload to google.oauth2.credentials.Credentials."""
    from google.oauth2.credentials import Credentials

    # Do NOT pass scopes= here. Tauri grants broader scopes (gmail.modify,
    # calendar.events) than the heartbeat's GOOGLE_SCOPES (readonly). If we
    # declare readonly scopes, google-auth may send them during auto-refresh,
    # which can trigger a "scope changed" error from Google. Omitting scopes
    # lets the Credentials object use whatever scopes were originally granted.
    return Credentials(  # type: ignore[no-untyped-call]
        token=payload.access_token,
        refresh_token=payload.refresh_token,
        token_uri="https://oauth2.googleapis.com/token",
        client_id=payload.client_id,
        client_secret=payload.client_secret,
    )


def load_credentials(
    engagement_id: str,
    token_path: Path,
    *,
    source: GoogleSource = "calendar",
) -> Credentials:
    """Load Google credentials, trying Firestore first, then legacy file.

    This is the Phase F replacement for ``load_google_credentials``.
    Callers should migrate from:
        ``load_google_credentials(token_path, source=source)``
    to:
        ``load_credentials(engagement_id, token_path, source=source)``
    """

    # Try Firestore path first
    try:
        result = _read_firestore_token(engagement_id, source)
    except GoogleAuthFailure:
        raise
    except Exception as exc:
        logger.warning("Firestore token read failed (falling back to file): %s", exc)
        result = None

    if result is not None:
        payload, doc = result

        if _is_expired(payload):
            if not payload.refresh_token:
                raise GoogleAuthFailure(
                    CollectorError(
                        source=source,
                        error_code="oauth_refresh_failed",
                        message=(
                            "Firestore token has no refresh_token. "
                            "Re-authenticate in the Tauri app."
                        ),
                    )
                )
            try:
                payload = _refresh_and_writeback(payload, engagement_id, doc)
            except Exception as exc:
                raise GoogleAuthFailure(
                    CollectorError(
                        source=source,
                        error_code="oauth_refresh_failed",
                        message=f"Token refresh failed: {exc}. Re-authenticate in the Tauri app.",
                    )
                ) from exc

        return _payload_to_credentials(payload)

    # Fallback to legacy file-based path
    logger.info("No Firestore token for engagement %s; falling back to token file", engagement_id)
    return load_google_credentials(token_path, source=source)

File: src/lib/firestore-tokens.ts
/**
 * Phase F — Encrypt OAuth tokens and sync to Firestore.
 *
 * After a successful per-engagement Google OAuth, the token payload
 * (already stored in the Mac keychain by redirect_server.rs) is
 * encrypted with AES-256-GCM using an operator-supplied key and
 * written to engagements/{eid}/google_tokens/google in Firestore.
 *
 * The heartbeat (Admin SDK) reads + decrypts per-engagement.
 * Refresh-token rotation writes back encrypted.
 *
 * Encryption uses the WebCrypto API (available in Tauri's webview).
 * The ciphertext field contains ciphertext || 16-byte GCM auth tag
 * (WebCrypto's default output format).
 */

import { doc, getDoc, setDoc, serverTimestamp, updateDoc } from "firebase/firestore";
import { db } from "./firebase";

const KEY_ENV = import.meta.env.VITE_TOKEN_ENCRYPTION_KEY ?? "";
const KEY_VERSION = Number(import.meta.env.VITE_TOKEN_ENCRYPTION_KEY_VERSION ?? "1");

/**
 * Decode the base64-encoded 32-byte AES key from the env var.
 * Returns null if the key is missing or malformed.
 */
function getEncryptionKey(): Uint8Array | null {
  if (!KEY_ENV) return null;
  try {
    const raw = Uint8Array.from(atob(KEY_ENV), (c) => c.charCodeAt(0));
    if (raw.length !== 32) return null;
    return raw;
  } catch {
    return null;
  }
}

/**
 * Encrypt a plaintext string with AES-256-GCM.
 * Returns { ciphertext, iv } both as base64 strings.
 * ciphertext includes the 16-byte GCM auth tag appended by WebCrypto.
 */
async function encrypt(
  plaintext: string,
  keyBytes: Uint8Array,
): Promise<{ ciphertext: string; iv: string }> {
  const key = await crypto.subtle.importKey(
    "raw",
    keyBytes,
    { name: "AES-GCM" },
    false,
    ["encrypt"],
  );

  // 12-byte random IV — never reuse with the same key.
  const iv = crypto.getRandomValues(new Uint8Array(12));

  const encoded = new TextEncoder().encode(plaintext);
  const encrypted = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv },
    key,
    encoded,
  );

  return {
    ciphertext: uint8ToBase64(new Uint8Array(encrypted)),
    iv: uint8ToBase64(iv),
  };
}

function uint8ToBase64(bytes: Uint8Array): string {
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary);
}

/**
 * Encrypt the token payload and write it to Firestore.
 *
 * Called from SettingsView / ChatView after oauth:token-stored fires.
 * The token payload is read from the keychain via getCredential().
 *
 * THROWS if the encryption key is missing. The caller should surface
 * this to the user — a missing key means Firestore sync is broken
 * and the heartbeat will never see the new token.
 */
export async function syncTokenToFirestore(
  engagementId: string,
  tokenPayloadJson: string,
): Promise<void> {
  const keyBytes = getEncryptionKey();
  if (!keyBytes) {
    throw new Error(
      "VITE_TOKEN_ENCRYPTION_KEY is missing or malformed in .env.local. " +
      "OAuth succeeded but the token was NOT synced to Firestore — " +
      "the heartbeat will not see this token. " +
      "Add the encryption key from the VM's /etc/ikrs-heartbeat/secrets.env " +
      "to .env.local and rebuild. See heartbeat/README.md.",
    );
  }

  const { ciphertext, iv } = await encrypt(tokenPayloadJson, keyBytes);

  // Fetch the connected email before writing — we'll store it as
  // an unencrypted field for UI display (not sensitive; just the
  // email address, not the token).
  let connectedEmail: string | null = null;
  try {
    const payload = JSON.parse(tokenPayloadJson);
    const resp = await fetch(
      "https://www.googleapis.com/oauth2/v3/userinfo",
      { headers: { Authorization: `Bearer ${payload.access_token}` } },
    );
    if (resp.ok) {
      const info = await resp.json();
      connectedEmail = info.email ?? null;
    }
  } catch {
    // Non-fatal — email will show as "unknown" in UI
  }

  const ref = doc(db, "engagements", engagementId, "google_tokens", "google");
  await setDoc(ref, {
    ciphertext,
    iv,
    keyVersion: KEY_VERSION,
    updatedAt: serverTimestamp(),
    writtenBy: "tauri",
    ...(connectedEmail ? { connectedEmail } : {}),
  });
}

/**
 * Read the cached connectedEmail from the google_tokens doc.
 * Returns null if not set or doc doesn't exist.
 */
export async function getConnectedEmail(
  engagementId: string,
): Promise<string | null> {
  const ref = doc(db, "engagements", engagementId, "google_tokens", "google");
  const snap = await getDoc(ref);
  if (!snap.exists()) return null;
  return (snap.data()?.connectedEmail as string) ?? null;
}

/**
 * Refresh the connectedEmail by calling userinfo with the current
 * access token from keychain, then updating the Firestore doc.
 */
export async function refreshConnectedEmail(
  engagementId: string,
  accessToken: string,
): Promise<string | null> {
  try {
    const resp = await fetch(
      "https://www.googleapis.com/oauth2/v3/userinfo",
      { headers: { Authorization: `Bearer ${accessToken}` } },
    );
    if (!resp.ok) return null;
    const info = await resp.json();
    const email = info.email as string | undefined;
    if (!email) return null;

    const ref = doc(db, "engagements", engagementId, "google_tokens", "google");
    await updateDoc(ref, { connectedEmail: email });
    return email;
  } catch {
    return null;
  }
}
